pkg/flow/internal/dag/dag.go

```diff
@@ -1,12 +1,28 @@
 // Package dag defines a Directed Acyclic Graph.
 package dag
 
-import "fmt"
+import (
+	"fmt"
+
+	"github.com/grafana/agent/pkg/river/ast"
+	"github.com/grafana/agent/pkg/river/vm"
+)
 
 // Node is an individual Vertex in the DAG.
 type Node interface {
```

Comment:
I'd like to keep the Node interface hyper-generic to leave the door open in the future for Nodes in the future which don't have blocks (i.e., #3029).

Instead, can you make a new interface in `pkg/flow/internal/controller` for this?

```go 
package controller 

import (
  "github.com/grafana/agent/pkg/flow/internal/dag"
)

// BlockNode is a node in the DAG which manages a River block 
// and can be evaluated.
type BlockNode interface {
	dag.Node 

	// Evaluate updates the arguments for the managed component
	// by re-evaluating its River block with the provided scope. The managed component
	// will be built the first time Evaluate is called.
	//
	// Evaluate will return an error if the River block cannot be evaluated or if
	// decoding to arguments fails.
	Evaluate(scope *vm.Scope) error

	// Block returns the current block of the managed config node.
	Block() *ast.BlockStmt
}
```

This means you may have to keep some of the type checks in, but instead you can type check based on the interface:

```go
switch cn := cn.(type) {
case BlockNode:
	traversals = expressionsFromBody(cn.Block().Body)
}
```